mmv1/third_party/terraform/provider/provider_billing_project_test.go

```diff
@@ -0,0 +1,326 @@
+				),
+			},
+			{
+				// GOOGLE_CLOUD_QUOTA_PROJECT is NOT used here
+				PreConfig: func() {
+					t.Setenv("GOOGLE_BILLING_PROJECT", "")
+					t.Setenv("GOOGLE_CLOUD_QUOTA_PROJECT", GOOGLE_CLOUD_QUOTA_PROJECT) // NOT used
+				},
+				Config: testAccSdkProvider_billing_project_inEnvsOnly(context),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr("data.google_provider_config_sdk.default", "billing_project", ""),
+				),
+			},
+		},
+	})
+}
+
+func testAccSdkProvider_billing_project_emptyStringValidation(t *testing.T) {
+	acctest.SkipIfVcr(t) // Test doesn't interact with API
+
+	billingProject := "my-billing-project-id"
+
+	// ensure all billing_project env vars set
+	t.Setenv("GOOGLE_BILLING_PROJECT", billingProject)
+
+	context := map[string]interface{}{
+		"billing_project": "", // empty string used
+	}
+
+	acctest.VcrTest(t, resource.TestCase{
+		// No PreCheck for checking ENVs
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccSdkProvider_billing_project_inProviderBlock(context),
+				PlanOnly:    true,
+				ExpectError: regexp.MustCompile("expected a non-empty string"),
+			},
+		},
+	})
+}
+
+func testAccSdkProvider_billing_project_useWithAndWithoutUserProjectOverride(t *testing.T) {
+
+	randomString := acctest.RandString(t, 10)
+	contextUserProjectOverrideFalse := map[string]interface{}{
+		"org_id":                envvar.GetTestOrgFromEnv(t),
+		"billing_account":       envvar.GetTestBillingAccountFromEnv(t),
+		"user_project_override": "false", // Used in combo with billing_account
+		"random_suffix":         randomString,
+	}
+
+	contextUserProjectOverrideTrue := map[string]interface{}{
+		"org_id":                envvar.GetTestOrgFromEnv(t),
+		"billing_account":       envvar.GetTestBillingAccountFromEnv(t),
+		"user_project_override": "true", // Used in combo with billing_account
+		"random_suffix":         randomString,
+	}
+
+	acctest.VcrTest(t, resource.TestCase{
+		// No PreCheck for checking ENVs
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				// Setup resources
+				// Neither user_project_override nor billing_project value used here
+				Config: testAccSdkProvider_billing_project_useBillingProject_setup(contextUserProjectOverrideFalse),
+			},
+			{
+				// With user_project_override=true the PubSub topic CANNOT be provisioned because quota is consumed
+				// from the newly provisioned project, and that project does not have the PubSub API enabled.
+				// The billing_project is used, leading to the error occurring, because user_project_override=true
+				Config:      testAccSdkProvider_billing_project_useBillingProject_scenario(contextUserProjectOverrideTrue),
+				ExpectError: regexp.MustCompile(fmt.Sprintf("Error 403: Cloud Pub/Sub API has not been used in project tf-test-%s", randomString)),
+			},
```

Comment:
[Test failure](https://github.com/GoogleCloudPlatform/magic-modules/pull/11610#issuecomment-2386175893) is because this step isn't failing as expected

Reply:
I'm unsure why this is the case. When I run this test locally it passed (i.e. there is an error created at the expected time). Hopefully running the tests again will give more insight?

Reply:
I'm seeing the following error
```
{
  "error": {
    "code": 400,
    "message": "The topic projects/<project>/topics/<topic> cannot be created or updated due to an organization policy specified at the project or organization level. The organization policy does not allow message storage in any GCP region where Cloud Pub/Sub is present. The organization policy may be updated by your project or organization administrator. See https://console.cloud.google.com/iam-admin/orgpolicies/gcp-resourceLocations?project=<project>. If the topic's project was recently created, you may need to wait a few minutes for the project's organization policy to be properly initialized, and then retry this operation.",
    "status": "FAILED_PRECONDITION"
  }
}
```

That I think is masking the expected 403 error?
I'm not sure why its only happening in the VCR runs and not the TC runs in the VCR test project.

Given that it works in TC/nightlies, I'm inclined to just skip in VCR for now but with a comment and monitor it

Reply:
Discussion from DMs:

- the "The organization policy does not allow... ...If the topic's project was recently created, you may need to wait a few minutes for the project's organization policy to be properly initialized, and then retry this operation" error on creating the Topic is retried multiple times
- eventually it succeeds: if you search for "2024/10/02 10:25:22 [DEBUG] Google API Request Details:" in the test logs  it'll take you to where the Topic is created successfully
- however it should _never_ be successful after the organization policy is properly initialized, as the API isn't active in the project being used


This is weird, so I'll try out changing which service/resource is used in this test. If that doesn't help I'll revert back to using PubSub and will just skip the test in VCR as this test is passing in TeamCity builds that use the VCR GCP project

Reply:
I tried swapping to using either a BigQuery dataset/Cloud KMS key ring/Compute Engine instance but they didn't have the expected behaviour when setting user_project_override was swapped to false - there was an error about the API not being enabled in the new project on step 3, when I'd expect step 3 to pass given that no project override is being set.

I think this is due to resource-oriented APIs versus client-oriented APIs. I _think_ this test needs the resource being used to belong to a client-oriented API, so the billing defaults to going to the project where the Terraform identity is, and being an API where it doesn't need to be enabled in _both_ projects to work.